src/resources/uart_esp32.cc

```diff
@@ -119,135 +608,230 @@ PRIMITIVE(init) {
+    tx_buffer_size = 1024;
+    rx_buffer_size = 2048;
+  } else if ((options & 4) != 0) {
+    // Medium speed
+    interrupt_flags |= ESP_INTR_FLAG_LEVEL2 | ESP_INTR_FLAG_LEVEL3;
+    full_interrupt_threshold = 92;
+    tx_buffer_size = 512;
+    rx_buffer_size = 1536;
+  } else {
+    // Low speed setting.
+    interrupt_flags |= ESP_INTR_FLAG_LEVEL1 | ESP_INTR_FLAG_LEVEL2 | ESP_INTR_FLAG_LEVEL3;
+    full_interrupt_threshold = 105;
+    tx_buffer_size = 256;
+    rx_buffer_size = 768;
+  }
 
-  uart_port_t port = kInvalidUartPort;
+  uart_initialization_record_t init = {};
+  uart_port_t port = determine_preferred_port(tx, rx, rts, cts);
 
-  // Check if there is a preferred device.
-  if ((tx == -1 || tx == 17) &&
-      (rx == -1 || rx == 16) &&
-      (rts == -1 || rts == 7) &&
-      (cts == -1 || cts == 8)) {
-    port = UART_PORT;
-  }
-  if ((tx == -1 || tx == 10) &&
-      (rx == -1 || rx == 9) &&
-      (rts == -1 || rts == 11) &&
-      (cts == -1 || cts == 6)) {
-    port = UART_NUM_1;
-  }
-  if ((tx == -1 || tx == 1) &&
-      (rx == -1 || rx == 3) &&
-      (rts == -1 || rts == 22) &&
-      (cts == -1 || cts == 19)) {
-    port = UART_NUM_0;
-  }
   port = uart_ports.preferred(port);
   if (port == kInvalidUartPort) OUT_OF_RANGE;
+  init.port = port;
 
   ByteArray* proxy = process->object_heap()->allocate_proxy();
   if (proxy == null) {
-    uart_ports.put(port);
+    uart_init_cleanup(init);
     ALLOCATION_FAILED;
   }
 
-  int flow_ctrl = 0;
-  if (mode == UART_MODE_UART) {
-    if (rts != -1) flow_ctrl += UART_HW_FLOWCTRL_RTS;
-    if (cts != -1) flow_ctrl += UART_HW_FLOWCTRL_CTS;
+  init.queue = xQueueCreate(EVENT_QUEUE_SIZE, sizeof(uart_event_types_t));
+  if (!init.queue) {
+    uart_init_cleanup(init);
+    MALLOC_FAILED;
   }
 
-  uart_config_t uart_config = {
-    .baud_rate = baud_rate,
-    .data_bits = (uart_word_length_t)(data_bits - 5),
-    .parity = UART_PARITY_DISABLE,
-    .stop_bits = UART_STOP_BITS_1,
-    .flow_ctrl = (uart_hw_flowcontrol_t)flow_ctrl,
-    .rx_flow_ctrl_thresh = 122,
-    .source_clk = UART_SCLK_APB,
-  };
+  init.hal = uart_toit_hal_init(port);
+  if (!init.hal) {
+    uart_init_cleanup(init);
+    MALLOC_FAILED;
+  }
 
+  init.rx_buffer = static_cast<uint8_t*>(malloc(rx_buffer_size));
+  if (!init.rx_buffer) {
+    uart_init_cleanup(init);
+    MALLOC_FAILED;
+  }
 
-  switch (stop_bits) {
-    case 2: uart_config.stop_bits = UART_STOP_BITS_1_5; break;
-    case 3: uart_config.stop_bits = UART_STOP_BITS_2; break;
+  init.tx_buffer = static_cast<uint8_t*>(malloc(tx_buffer_size));
+  if (!init.tx_buffer) {
+    uart_init_cleanup(init);
+    MALLOC_FAILED;
+  }
+
+  init.uart_resource = _new UartResource(group, port, init.queue, init.hal,
+                                         init.rx_buffer, rx_buffer_size,
+                                         init.tx_buffer, tx_buffer_size);
+  if (!init.uart_resource) {
+    uart_init_cleanup(init);
+    MALLOC_FAILED;
```

Comment:
Here you would add a call to the init object that flips a flag so that the destructor does nothing if called after this point.